subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c

```diff
@@ -218,52 +214,46 @@ allocate_output_surface (GstMsdkDec * thiz)
         break;
     }
   }
-#ifndef _WIN32
-  if ((gst_buffer_pool_acquire_buffer (thiz->alloc_pool, &out_buffer, NULL))
-      != GST_FLOW_OK) {
-    GST_ERROR_OBJECT (thiz, "Failed to allocate output buffer");
-    return NULL;
-  }
 
-  mem = gst_buffer_peek_memory (out_buffer, 0);
-  msdk_surface = g_slice_new0 (GstMsdkSurface);
-  msdk_surface->from_qdata = FALSE;
+  if (thiz->use_video_memory) {
+    if ((gst_buffer_pool_acquire_buffer (thiz->alloc_pool, &out_buffer, NULL))
+        != GST_FLOW_OK) {
+      GST_ERROR_OBJECT (thiz, "Failed to allocate output buffer");
+      return NULL;
+    }
 
-  if ((mfx_surface = gst_mini_object_get_qdata (GST_MINI_OBJECT_CAST (mem),
-              GST_MSDK_FRAME_SURFACE))) {
-    msdk_surface->surface = mfx_surface;
-    msdk_surface->from_qdata = TRUE;
+    msdk_surface =
+        gst_msdk_import_to_msdk_surface (out_buffer, thiz->context, NULL,
+        GST_MAP_WRITE);
   } else {
-    GST_ERROR ("No available surfaces");
-    g_slice_free (GstMsdkSurface, msdk_surface);
-    return NULL;
-  }
-#else
-  if (!gst_buffer_pool_is_active (thiz->pool) &&
-      !gst_buffer_pool_set_active (thiz->pool, TRUE)) {
-    GST_ERROR_OBJECT (thiz, "Failed to activate buffer pool");
-    return NULL;
-  }
+    if (!gst_buffer_pool_is_active (thiz->pool) &&
```

Comment:
Actually, thiz->pool is a video memory pool, right? So you cannot use gst_msdk_import_sys_mem_to_msk_surface...
 In decide_allocation, if the downstream offered pool is not available, we will use thiz->pool as our thiz->alloc_pool, so in this "else" path, we should not use thiz->pool to alloc buffer. Besides, if we want to handle the condition thiz->use_video_memory is FALSE, we should use thiz->other_pool and thiz->other_pool only exists when do_copy flag is true. 
Anyway, I think if we insist to use video memory for the decoding process, we don't need to decide if thiz->use_video_memory as true in this function.